### ISSUE TITLE
[FeatureRequest] Gather sparse logprobs 

### ISSUE THREAD

**Issue by Marks101** [NVIDIA/TensorRT-LLM#1419]
Hello team,

We typically use `gather_all_token_logits` to collect the logit tensors for post-processing. Especially for large vocabulary sizes (128 000) this can require a lot of GPU memory. For example, when running inference loads with input and output lengths of 1024 and a batch size of 32, the collected logit tensor requires 32 GB of memory (fp32).

In vllm it is possible to collect only the topk logprobs (see [here](https://github.com/vllm-project/vllm/blob/0ce0539d4750f9ebcd9b19d7085ca3b934b9ec67/vllm/sampling_params.py#L85)). This is much more memory efficient and would be sufficient for our purposes. Is there currently a way to do this in TensorRT-LLM as well? If not, we would really appreciate this feature in both `ModelRunner` and `ModelRunnerCpp`.

This issue is somehow related to https://github.com/NVIDIA/TensorRT-LLM/issues/1040, as it would be possible to solve it on our side if it is possible to collect arbitrary model outputs.

Thank you

**Comment by Marks101** [comment]
@byshiue @ncomly-nvidia we figured that this feature could be implemented on our side based on a `LogitsProcessor`. But currently these are not supported by the `ModelRunnerCpp` / `tensorrt_llm.bindings.GptSession`:
https://github.com/NVIDIA/TensorRT-LLM/blob/71d8d4d3dc655671f32535d6d2b60cab87f36e87/tensorrt_llm/runtime/model_runner_cpp.py#L310-L312
Is there any plan to extend the support?

**Comment by MartinMarciniszyn** [comment]
@Marks101, the logits processor is supported on `ModelRunnerCppExecutor`:
https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/runtime/model_runner_cpp.py#L48
Could you try that please?

**Comment by Marks101** [comment]
Hi @MartinMarciniszyn thank you for the update. We will take a look at this 😃

**Comment by shangshng** [comment]
> @Marks101, the logits processor is supported on `ModelRunnerCppExecutor`: https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/runtime/model_runner_cpp.py#L48 Could you try that please?

Hello, it looks like logits processor is disabled here....
![image](https://github.com/NVIDIA/TensorRT-LLM/assets/22979596/5532c0ed-12fe-42a7-9985-0c7f23e1e0aa)

**Comment by MartinMarciniszyn** [comment]
Thanks for the feedback @shangshng. It should be support in the Python bindings of the Executor API. @dcampora, could you please add support to `ModelRunnerCpp`?

@Marks101, you can use the Executor API directly instead of going through `ModelRunnerCpp` for now.
